### PR TITLE
feature: support for StartPreserveBlock and EndPreserveBlock instruct…

### PR DESCRIPTION
**Pull request by sergei-kag**
…ions.

*Issue #, if available:*

*Description of changes:*
Support added for StartPreserveBlock and EndPreserveBlock instructions. The code between the start and end instructions will not be optimized by removing, reordering, or decomposing. Thus, creating a preserved block of code.

I've also updated the error messages in program_v1.py. There are three code paths that all had the same error message, so I've separated them into unique messages to help with future debugging.

*Testing done:*
```
==================================================================================== test session starts =====================================================================================
platform darwin -- Python 3.7.6, pytest-6.2.4, py-1.10.0, pluggy-0.13.1 -- /usr/local/opt/python/bin/python3.7
cachedir: .pytest_cache
rootdir: .../amazon-braket-schemas-python, configfile: setup.cfg
collected 14 items                                                                                                                                                                           

test/unit_tests/braket/ir/jaqcd/test_program_v1.py::test_missing_instructions_property XFAIL                                                                                           [  7%]
test/unit_tests/braket/ir/jaqcd/test_program_v1.py::test_non_instruction XFAIL                                                                                                         [ 14%]
test/unit_tests/braket/ir/jaqcd/test_program_v1.py::test_wrong_instruction XFAIL                                                                                                       [ 21%]
test/unit_tests/braket/ir/jaqcd/test_program_v1.py::test_wrong_compiler_directive XFAIL                                                                                                [ 28%]
test/unit_tests/braket/ir/jaqcd/test_program_v1.py::test_partial_non_instruction XFAIL                                                                                                 [ 35%]
test/unit_tests/braket/ir/jaqcd/test_program_v1.py::test_partial_non_result XFAIL                                                                                                      [ 42%]
test/unit_tests/braket/ir/jaqcd/test_program_v1.py::test_instruction_no_results PASSED                                                                                                 [ 50%]
test/unit_tests/braket/ir/jaqcd/test_program_v1.py::test_instruction_no_noise_results PASSED                                                                                           [ 57%]
test/unit_tests/braket/ir/jaqcd/test_program_v1.py::test_instruction_with_results PASSED                                                                                               [ 64%]
test/unit_tests/braket/ir/jaqcd/test_program_v1.py::test_partial_non_rotation_basis_instruction XFAIL                                                                                  [ 71%]
test/unit_tests/braket/ir/jaqcd/test_program_v1.py::test_no_rotation_basis_instruction PASSED                                                                                          [ 78%]
test/unit_tests/braket/ir/jaqcd/test_program_v1.py::test_rotation_basis_instruction PASSED                                                                                             [ 85%]
test/unit_tests/braket/ir/jaqcd/test_program_v1.py::test_start_preserve_block_instruction PASSED                                                                                       [ 92%]
test/unit_tests/braket/ir/jaqcd/test_program_v1.py::test_end_preserve_block_instruction PASSED                                                                                         [100%]

================================================================================ 7 passed, 7 xfailed in 0.19s ================================================================================
~
```
## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/amazon-braket-schemas-python/blob/main/CONTRIBUTING.md) doc
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/amazon-braket-schemas-python/blob/main/CONTRIBUTING.md#commit-your-change)
- [ ] I have updated any necessary documentation, including [READMEs](https://github.com/aws/amazon-braket-schemas-python/blob/main/README.md) and [API docs](https://github.com/aws/amazon-braket-schemas-python/blob/main/CONTRIBUTING.md#documentation-guidelines) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/braket/ir/jaqcd/instructions.py [hunk 1]
```diff
@@ -899,3 +900,40 @@ class Type(str, Enum):
         kraus = "kraus"
 
     type = Type.kraus
+
+    """
+    StartPreserveBLock.
```
Comment (jheckey) [inline src/braket/ir/jaqcd/instructions.py:905]: Please provide a description of the functionality.

File: src/braket/ir/jaqcd/instructions.py [hunk 2]
```diff
@@ -899,3 +900,40 @@ class Type(str, Enum):
...
+
+    class Type(str, Enum):
+        start_preserve_block = "start_preserve_block"
+
+    type = Type.start_preserve_block
+
+
+class EndPreserveBlock(CompilerDirective):
+    """
+    EndPreserveBLock.
```
Comment (jheckey) [inline src/braket/ir/jaqcd/instructions.py:925]: Please provide a description of the functionality.

File: src/braket/ir/jaqcd/shared_models.py [hunk 3]
```diff
@@ -335,3 +335,18 @@ class MultiState(BaseModel):
...
 
     states: conlist(constr(regex="^[01]+$", min_length=1), min_items=1)
+
+class CompilerDirective(BaseModel):
+    """
+    A Compiler Directive to preserve a block of code between StartPreserveBlock
+    and EndPreserveBlock directives.
+
+    Attributes:
+        directive (List [Start PreserveBlock | EndPreserveBlock])
```
Comment (jheckey) [inline src/braket/ir/jaqcd/shared_models.py:345]: Typo: `Start PreserveBlock` -> `StartPreserveBlock`

**Review by jheckey — CHANGES_REQUESTED** [review]

File: src/braket/ir/jaqcd/shared_models.py (hunk 3 above)
Comment (sergei-kag) [reply src/braket/ir/jaqcd/shared_models.py]: Updated

File: src/braket/ir/jaqcd/instructions.py (hunk 2 above)
Comment (sergei-kag) [reply src/braket/ir/jaqcd/instructions.py]: Updated

File: src/braket/ir/jaqcd/instructions.py (hunk 1 above)
Comment (sergei-kag) [reply src/braket/ir/jaqcd/instructions.py]: Updated

File: src/braket/ir/jaqcd/instructions.py [hunk 4]
```diff
@@ -899,3 +900,43 @@ class Type(str, Enum):
         kraus = "kraus"
 
     type = Type.kraus
+
+
+class StartPreserveBlock(CompilerDirective):
+    """
+    StartPreserveBLock is a compiler instruction to start the code block which
+    will preserve the instruction within StartPreserveBlock and EndPreserveBlock
+    from being removed, reordered, or decompositioned.
```
Comment (jheckey) [inline src/braket/ir/jaqcd/instructions.py:909]: Rewording: "being modified in anyway by the compiler."

File: src/braket/ir/jaqcd/program_v1.py [hunk 5]
```diff
@@ -216,6 +227,8 @@ def validate_instructions(cls, value, field):
                 return _valid_gates[value["type"]](**value)
             elif value["type"] in _valid_noise_channels:
                 return _valid_noise_channels[value["type"]](**value)
+            elif value["type"] in _valid_compiler_directives:
+                return _valid_compiler_directives[value["type"]](**value)
```
Comment (jheckey) [inline src/braket/ir/jaqcd/program_v1.py:231]: Annoying that the coverage still isn't getting there.

**Review by jheckey — CHANGES_REQUESTED** [review]

File: src/braket/ir/jaqcd/instructions.py (hunk 4 above)
Comment (sergei-kag) [reply src/braket/ir/jaqcd/instructions.py]: Updated

File: src/braket/ir/jaqcd/program_v1.py (hunk 5 above)
Comment (sergei-kag) [reply src/braket/ir/jaqcd/program_v1.py:231]: Updated tests to get 100% coverage.

**Review by jheckey — APPROVED** [review]

File: src/braket/ir/jaqcd/instructions.py [hunk 6]
```diff
@@ -899,3 +900,43 @@ class Type(str, Enum):
...
+    EndPreserveBLock is a compiler instruction to mark the end of the code block
+    which will preserve the instruction within StartPreserveBlock and EndPreserveBlock
+    from being  modified in any way by the compiler.
+
+    Attributes:
+        type (str): The instruction type. default = "end_preserve_block". (type) is optional.
+            This should be unique among all instruction types.
+
+    Examples:
+        >>> EndPreserveBlock(directive=["EndPreserveBlock"])
```
Comment (kshitijc) [inline src/braket/ir/jaqcd/instructions.py:936]: This seems a little redundant. The information carried by the directive is the same as the class name and the type. Do we expect the directive to diverge from the instruction type ?

Comment (sergei-kag) [reply src/braket/ir/jaqcd/instructions.py]: Introduced the default directive for StartVerbatimBlock and EndVerbatimBlock classes to reduce the redundancy.

**Review by jheckey — APPROVED** [review]

File: src/braket/ir/jaqcd/instructions.py [hunk 7]
```diff
@@ -899,3 +900,47 @@ class Type(str, Enum):
...
+    Examples:
+        >>> EndVerbatimBlock()
+    """
+
+    class Type(str, Enum):
+        end_verbatim_block = "end_verbatim_block"
+
+    type = Type.end_verbatim_block
+
+    directive: str = "EndVerbatimBlock"
```
Comment (jheckey) [inline src/braket/ir/jaqcd/instructions.py:946]: Interestingly, you can use this to just do a copy-paste: `directive: str = __qualname__`

**Pull request merged**